errai-bus/src/main/java/org/jboss/errai/bus/client/framework/ClientMessageBusImpl.java

```diff
@@ -945,28 +945,53 @@ private void sendAllDeferred() {
     if (!deferredMessages.isEmpty())
       logger.info("transmitting deferred messages now ...");
 
+    List<Message> localMessages = new ArrayList<>();
+
     final List<Message> highPriority = new ArrayList<>();
     for (final Message message : new ArrayList<>(deferredMessages)) {
       if (message.hasPart(MessageParts.PriorityProcessing)) {
-        if (remotes.containsKey(message.getSubject()))
+        if (remotes.containsKey(message.getSubject())) {
           highPriority.add(message);
+        } else {
+          localMessages.add(message);
+        }
         deferredMessages.remove(message);
       }
     }
 
     final List<Message> lowPriority = new ArrayList<>();
     do {
       for (final Message message : new ArrayList<>(deferredMessages)) {
-        if (remotes.containsKey(message.getSubject()))
+        if (remotes.containsKey(message.getSubject())) {
           lowPriority.add(message);
+          localMessages.remove(message);
+        } else {
+          localMessages.add(message);
+        }
         deferredMessages.remove(message);
       }
     }
     while (!deferredMessages.isEmpty());
 
-    transportHandler.transmit(highPriority);
-    transportHandler.transmit(lowPriority);
-    deferredMessages.clear();
+    try {
+      transportHandler.transmit(highPriority);
+      transportHandler.transmit(lowPriority);
+
+      for (final Message message : localMessages) {
+        String subject = message.getSubject();
+        if(!localSubscriptions.containsKey(subject)) {
```

Comment:
This is almost there, but I think we need to change this condition to this:

    !(localSubscriptions.containsKey(subject) 
      || subscriptions.containsKey(subject)
      || shadowSubscriptions.containsKey(subject))

If you look in the implementation of the send method in the bus. Those three fields + remotes are all the places where a message can be delivered, so if any of those three contains a matching subject then the message had subscribers.

Reply:
Ah okay, wasn't sure if they were all required. Thanks! Will make the changes now.